core/amber/src/main/python/core/runnables/test_batch_operator.py

```diff
@@ -0,0 +1,861 @@
+        )
+        return ControlElement(tag=mock_controller, payload=payload)
+
+    @pytest.fixture
+    def data_processor(self, input_queue, output_queue, mock_udf, mock_link):
+        data_processor = DataProcessor(input_queue, output_queue)
+        # mock the operator binding
+        data_processor._operator = mock_udf
+        data_processor.context.batch_to_tuple_converter.update_all_upstream_link_ids(
+            {mock_link}
+        )
+        data_processor._operator.output_schema = {
+            "test-1": "string",
+            "test-2": "integer",
+        }
+        yield data_processor
+        data_processor.stop()
+
+    @pytest.fixture
+    def dp_thread(self, data_processor, reraise):
+        def wrapper():
+            with reraise:
+                data_processor.run()
+
+        dp_thread = Thread(target=wrapper, name="dp_thread_2")
+        yield dp_thread
+
+    @pytest.mark.timeout(2)
+    def test_dp_thread_can_start(self, dp_thread):
+        dp_thread.start()
+        assert dp_thread.is_alive()
+
+    @pytest.mark.timeout(5)
+    def test_dp_thread_can_process_batch_simple(
+            self,
+            mock_udf,
+            mock_link,
+            mock_receiver_actor,
+            mock_controller,
+            input_queue,
+            output_queue,
+            mock_data_elements,
+            dp_thread,
+            mock_update_input_linking,
+            mock_add_partitioning,
+            mock_end_of_upstream,
+            mock_query_statistics,
+            mock_batch,
+            command_sequence,
+            reraise,
+    ):
+        dp_thread.start()
+
+        # can process UpdateInputLinking
+        input_queue.put(mock_update_input_linking)
+
+        assert output_queue.get() == ControlElement(
+            tag=mock_controller,
+            payload=ControlPayloadV2(
+                return_invocation=ReturnInvocationV2(
+                    original_command_id=command_sequence,
+                    control_return=ControlReturnV2(),
+                )
+            ),
+        )
+
+        # can process AddPartitioning
+        input_queue.put(mock_add_partitioning)
+        assert output_queue.get() == ControlElement(
+            tag=mock_controller,
+            payload=ControlPayloadV2(
+                return_invocation=ReturnInvocationV2(
+                    original_command_id=command_sequence,
+                    control_return=ControlReturnV2(),
+                )
+            ),
+        )
+
+        # can process a InputDataFrame
+        for i in range(50):
+            input_queue.put(mock_data_elements[i])
+
+        output_data_element = []
+        for i in range(50):
+            output_data_element.append(output_queue.get())
+
+        input_queue.put(mock_query_statistics)
+        assert output_data_element[0].tag == mock_receiver_actor
+        assert isinstance(output_data_element[0].payload, OutputDataFrame)
```

Comment:
assert data elements before you input another control element.

make sure you test like this sequence:
1. add some input 
2. wait for some output
3. assert/check the output and states
4. repeat from 1 for next test input.

Do not overlap them.

Reply:
fix